modules/fvad/fvad.c

```diff
@@ -0,0 +1,348 @@
+}
+
+
+static void print_vad(int pos, int color, bool tx, bool active)
+{
+	/* move cursor to a fixed position */
+	re_fprintf(stderr, "\x1b[%dG", pos);
+
+	/* print vad in Nice colors */
+	re_fprintf(stderr, " \x1b[%dm[%s]\x1b[;m\r",
+		   color,  active ? (tx ? "tx" : "rx") : "  ");
+}
+
+
+static void find_first_call(struct call *call, void *arg)
+{
+	struct filter_arg *fa = arg;
+
+	if (!fa->call)
+		fa->call = call;
+}
+
+
+static bool find_call(const struct call *call, void *arg)
+{
+	struct filter_arg *fa = arg;
+
+	return call_audio(call) == fa->audio;
+}
+
+
+static int encode_update(struct aufilt_enc_st **stp, void **ctx,
+			 const struct aufilt *af, struct aufilt_prm *prm,
+			 const struct audio *au)
+{
+	struct vad_enc *st;
+	struct filter_arg fa = { au, NULL };
+	(void)ctx;
+
+	if (!stp || !af || !prm)
+		return EINVAL;
+
+	if (*stp)
+		return 0;
+
+	if (prm->ch != 1)
+		return EINVAL;
+
+	st = mem_zalloc(sizeof(*st), enc_destructor);
+	if (!st)
+		return ENOMEM;
+
+	st->fvad = fvad_new();
+	if (!st->fvad) {
+		mem_deref(st);
+		return ENOMEM;
+	}
+
+	int err = fvad_set_sample_rate(st->fvad, prm->srate);
+	if (err < 0) {
+		mem_deref(st);
+		return EINVAL;
+	}
+
+	uag_filter_calls(find_first_call, find_call, &fa);
+
+	*stp = (struct aufilt_enc_st *)st;
+
+	st->call = fa.call;
+
+	return 0;
+}
+
+
+static int decode_update(struct aufilt_dec_st **stp, void **ctx,
+			 const struct aufilt *af, struct aufilt_prm *prm,
+			 const struct audio *au)
+{
+	struct vad_dec *st;
+	struct filter_arg fa = { au, NULL };
+	(void)ctx;
+
+	if (!stp || !af || !prm)
+		return EINVAL;
+
+	if (*stp)
+		return 0;
+
+	if (prm->ch != 1)
+		return EINVAL;
+
+	st = mem_zalloc(sizeof(*st), dec_destructor);
+	if (!st)
+		return ENOMEM;
+
+	st->fvad = fvad_new();
+	if (!st->fvad) {
+		mem_deref(st);
+		return ENOMEM;
+	}
+
+	int err = fvad_set_sample_rate(st->fvad, prm->srate);
+	if (err < 0) {
+		mem_deref(st);
+		return EINVAL;
+	}
+
+	uag_filter_calls(find_first_call, find_call, &fa);
+
+	st->call = fa.call;
+
+	*stp = (struct aufilt_dec_st *)st;
+
+	return 0;
+}
+
+
+static bool auframe_vad(Fvad *fvad, struct auframe *af)
+{
+	static int chunk_times_ms[] = { 30, 20, 10 };
+
+	int16_t *buf = NULL;
+	int16_t *allocated = NULL;
+	bool detected = false;
+
+	/* convert to 16 bit linear */
+	if (af->fmt == AUFMT_S16LE) {
+		buf = af->sampv;
+	}
+	else {
+		allocated = (int16_t*)mem_alloc(sizeof(int16_t) * af->sampc,
+			NULL);
+		buf = allocated;
+		auconv_to_s16(buf, af->fmt, af->sampv, af->sampc);
+	}
+
+	size_t ms = af->sampc * 1000 / af->srate;
+	size_t pos = 0;
+
+	/* process all chunk_sizes that fvad accepts */
+	for (size_t chunk_time_index = 0;
+		chunk_time_index < RE_ARRAY_SIZE(chunk_times_ms);
+		++chunk_time_index) {
+
+		const size_t chunk_time = chunk_times_ms[chunk_time_index];
+		const size_t sampc = af->srate / 1000 * chunk_time;
+
+		for (size_t i = 0; i < ms / chunk_time; ++i) {
+			int err = fvad_process(fvad, buf + pos, sampc);
```

Comment:
not sure if this is correct

If `af->srate` is 16000 and `chunk_time` is 60, then
- `sampc == 480` but
- `af->sampc == 320`

The function `fvad_process()` reads to far. Please solve this and check with valgrind that there is no illegal read!
-